[tutorials] Working with Lists tutorial (#730)
lit/lit.dev: packages/lit-dev-content/site/tutorials/content/working-with-lists/00.md

> In this tutorial, we will be exploring different ways of rendering a list of

Reviewer: no we:

> This tutorial will be exploring...

items in Lit!

> Let's begin by looking a how we might render a list of names from an array.

Reviewer: Let's not use` Let's` who is the implied `we`? is it the Lit team, the user and us, is it the Lit library itself?

Suggestion:

> Begin my rendering a list of names from an array. The example already has set up a custom element with the `names` property that is an array of strings.

some help:

- commands: just give the command
  - lets begin -> begin
  - we can see -> "See" or "You may notice" or plainly "Notice"
    - prefer the assertive voice rather than "may" so the user can be assured they know that we know what we're teaching them
- we
  - tell the user who we is
  - we -> the example / the code

We've already set up a custom element with a `names` property that is an
array of strings.

> We can utlize the `map` array prototype method to generate a list item

Reviewer: Fom here i'll just leave comments saying "we" or "lets" so you can keep track on where this stuff is

> We can utlize the `map` array prototype method to generate a list item

Reviewer: Perhaps before going into the first action for the user add a section on "prior knowledge: docs link docs link docs link"

like reactive properties. We expect you to know how reactive properties work and handle updates

Author: Good idea. How about also adding like a recommend prerequisite tutorial if applicable?

Something like
> Prior knowledge: [Defining a custom element, rendering a template, and reactive properties](http://localhost:5415/docs/components/overview/).
> Recommended prerequisite: [Intro to Lit](https://lit.dev/tutorials/intro-to-lit/).

to be placed under the introductory description of the tutorial.

Reviewer: That's the idea! Hopefully once we get a bunch of tutorials up i can go through and link to each other

for each item in the array within the unordered list (`<ul></ul>`).

> ```ts

Reviewer: Tell the user directly to add this too the code and add the filename before the sample:
`### my-element.ts`

Author: As I'm thinking about this change, I'm wondering whether this much spoon feeding is the best way to write interactive tutorials. In my former teaching experience, I'd never want to give away the answer but provide enough clues for students to figure out themselves which makes it more memorable. I know during the kickoff you mentioned people are lazy and want braindead solutions but in that case they can just click the solve button to see the solution. And if they really want to just see working code, they can get that in the doc section or the completed playground examples.

Reviewer: It might be doable skipping that for this tutorial since there's only one JS file and one HTML file. Other tutorials have multiple files and due to an oddity on how playground-ide works, it sometimes points to weird files

Author: Oh, I'm happy to add filenames for snippets but I was more getting at not providing copy-pasteable code. Prompt can say use the `map` prototype method to render an array of lit-html templates that wrap each item of the array in a list item tag, and the reader has to come up with the code based on that. Maybe I'm thinking too much like a learn-how-to-code classroom kind of tutorial where making the learners think seem more important.

Reviewer: also works. As long as it's glance-able

Reviewer 2: I think this is a really interesting idea. I think Tour of Go, for example, does things like this--poses a problem and lets you solve it. And I think we could do that--especially with the addition of the Check button. 

It might make sense, for example, to add a separate "Exercise" step (or steps in a longer tutorial) where you give general instructions and let the users work out a solution. (With the backup of the Check/Solve buttons.) 

Another way to do this is by adding suggestions--typically at the end of a step or the end of a tutorial--of things the user could try. For example, the "Extra credit" items at the end of Steve's tooltip tutorial: https://pr740-77daa6d---lit-dev-5ftespv5na-uc.a.run.app/tutorials/tooltip?mods=tutorialCatalog#09

Encouraging people to play with the code + giving that extra reward of figuring it out yourself. As you said, it really helps make the information stick.

> This tutorial will be exploring different ways of rendering a list of items in

Reviewer 3: Do we want to say "different ways of rendering a list" or "different ways of rendering lists"?

You can render more than one list at once, so I'd lean towards plural. @arthurevans ?

Lit!

> Lit makes it easy to render any lists of items. It has built-it support for
> iterables of any supported values. This, along with Lit's ability to render
> nested templates and helpful built-in directives, allow for rendering any
> variety of lists.

Reviewer 3: ```suggestion
Lit makes it easy to render lists of items, with has built-it support for
iterables: pass an array (or other iterable) to a child expression and Lit will render each item in the array. This, combined with Lit's ability to render nested templates, and JavaScript patterns for transforming lists of data into lists of templates, allows for very flexible list handling.
```

^ Not sure if this is too complicated or not, but I thought it might summarize list support quickly.

## What you'll learn
- Utilizing the `map` directive
- Using array prototype methods

> - Building static list and using the `range` directive

Reviewer 3: ```suggestion
- Building static lists and using the `range` directive
```

> - Building static list and using the `range` directive

Reviewer 3: The term "static list" might not be immediately apparent. Maybe "Generating lists with the range directive"

- Imperatively building an array of items to render

> - Utilizing the `repeat` directive

Reviewer 3: "Utilizing" is repeated here. I wonder if we want to add more description of what readers will do with each thing, like:

```
 - Rendering lists with the map() directive
 - Transforming arrays into templates with array.map()
 - Generating lists with the range() directive
 - Imperatively building arrays of items
 - Diffing lists of stateful elements with the repeat() directive
 - Using event listeners in a list

Author: @arthurevans do you have thoughts on whether we should add the `()` after the directive names? Scanning through https://lit.dev/docs/templates/directives/ has examples of both with and without the parens.

Reviewer 2: That's a good question. Originally I thought we were going to go with parens after both method and directive names. The current state is inconsistent for both, unfortunately. I lean towards using the parens everywhere. But at any event, make sure they're consistent in your tutorial. I'll put this on the agenda for a future docs meeting.

Reviewer: I'm personally against parens even for methods because the arguments may be important and it'd be kinda silly to include the method signature when simply referring to it

Reviewer 2: Yeah, using empty parens is kind of convention for marking, "this is a function/method". It's recommended by the Google developer docs style guide, which we approximately follow:

https://developers.google.com/style/code-in-text#methods

Back in the polymer days, I bucked this trend because I'd worked on Java interfaces in the past and the JavaDoc rule is that you don't include empty parens--because of the way Java function overloading works, `foo()` may literally be a different function than `foo(bar)`. 

At present, it seems like like the JS world is divided on the value of the parens. React docs appear to use them: https://reactjs.org/docs/rendering-elements.html Vue docs don't. Svelte seems to use them for platform APIs like `event.preventDefault()`, but not for Svelte's own APIs, which seems odd. (Also, they prefix their own method names with a `.` to I guess show it's a member of an object, which also looks odd to me, so let's not do that.)

I don't have a super strong feeling about this one other than a general desire to fall in line with the developer docs style guide. But whatever we do, we should do it consistently.

Author: I can see arguments for both but lean towards @e111077's reasoning. I currently have consistently no parens within this tutorial and will keep it that way for now then.

Semi related to this, is do we have patterns for adding `.` in front of prototype methods like when referring to `.map`, `.filter` for arrays? For Step 3 of this tutorial, I've added the `.` if anything at least to make it stand out from the `map` directive.

Reviewer 2: We've never used that pattern that I'm aware of. It looks like it's just the one place you mention them, so I'd suggest saying something like: "the `filter` or `map` array methods". It seems to me that `.map` might be reasonably clear if the reader was aware you were trying to make the distinction between the map array method and the map directive. 

But the dot is easy to miss and not necessarily transparent to a more novice programmer. I'd suggest erring on the side of being a bit more explicit.

Author: Ooh I did not see that comment about it being in the Google developer docs style guide. I'd be happy to lean on that for this tutorial then. Thanks for the link! I'll also update the array method portion.

This tutorial will be exploring different ways of rendering a list of
items in Lit!

## Previous knowledge
It's recommended to have an understanding of
[defining a component](/docs/components/defining/),
[rendering](/docs/components/rendering/),and
[reactive properties](/docs/components/properties/) before
working on this tutorial.<br>

> Recommended prerequisite: [Intro to Lit](/tutorials/intro-to-lit/).

Reviewer 4: Nit: Found it odd (both language and format) to have a sentence above recommending certain docs, and then a non-sentence immediately below it recommending more. Maybe format as a bulleted list, or else just add the tutorial to the sentence?

Reviewer 5: It might also be nice to have a  really brief outline here to setup what's going to happen.

Author: Combined the "Previous knowledge" section into sentences, and added a bulleted list of "What you'll learn"

## Rendering a list

> Begin by rendering a list of names from an array. The example already has set

Reviewer 3: I think there's a bit of conceptual info that could be added here to make list support click for devs.

"List" support is really a combination of two things: lit-html supports nested template results as values, and it supports iterables of any supported value type. Combine those and you get iterables of template results.

It might be nice to start with something _very_ simple, like:

```ts
  render() {
    return html`
      <p>A list of my friends</p>
      ${['Chandler', 'Phoebe', 'Joey', 'Monica', 'Rachel', 'Ross']}
    `;
  }
```

And then say that you might want to include markup to you can do a bulleted list, so you can transform an array with `.map()`...

Author: Yes, that's a great suggestion! I did see that in the Lists doc section but didn't want to dedicate a whole step. Starting the example with that is a great way to bring it into the conversation.

> up a custom element with the `names` property that is an array of string.

Reviewer 4: ```suggestion
up a custom element with the `names` property that is an array of strings.
```
(or "of type string").

> The `map` array prototype method can be utilized to generate a list item

Reviewer 5: Def defer to @arthurevans here but I think we want to try to avoid passive voice and be more direct. Perhaps something like "Use Array's `map`  method to generate a list..."

Author: Elliott has pointed this out in other places too. Reworded.

for each item in the array. Place the code snippet below in the return of the
`render` method within the unordered list (`<ul></ul>`).

{% switchable-sample %}

```ts
// my-element.ts
${this.names.map((name) => html`<li>${name}</li>`)}
```

```js
// my-element.js

> ${this.names.map((name) => html`<li>${name}</li>`)}

Reviewer 5: A little more description of what's happening here would be great. In particular, this is leveraging Lit's (super power?) ability easily to compose templates.

Author: Good call, especially since this one's geared toward beginners. Added a note below the code snippet.
